### PR TITLE
Upgrade eslint

### PR DESCRIPTION
**Pull request by SpaceK33z**
**What kind of change does this PR introduce?**

Dep upgrade

**Did you add tests for your changes?**

N/A

**If relevant, link to documentation update:**

N/A

**Summary**

We used a very old version of eslint. This new version also makes converting to ES6 a LOT easier, since it has a `--fix` flag which can fix many ES6 stuff automatically. Note that for this to work, we'll need to add some extra rules to `.eslintrc`. This was out of the scope of this PR I think.

**Does this PR introduce a breaking change?**

No.

### CODE REVIEW

**Comment by Kovensky** [comment]
Making the eslint rules more specific (especially with eslint 3) should drastically reduce the chance I get caught in the lint check 😅

File: .eslintrc [hunk 1]
```diff
@@ -8,29 +8,41 @@
...
-		"no-console": 0,
-		"comma-dangle": 0,
-		"no-unexpected-multiline": 2,
-		"valid-jsdoc": 2
+		"strict": "off",
+		"camelcase": "off",
+		"curly": "off",
+		"indent": ["error", "tab", { "SwitchCase": 1 }],
+		"eol-last": "warn",
+		"no-shadow": "off",
```
Comment (Kovensky) [inline .eslintrc]: Also off by default

File: .eslintrc [hunk 2]
```diff
@@ -8,29 +8,41 @@
...
-		"no-inner-declarations": 1,
-		"no-loop-func": 1,
-		"space-before-function-paren": [2, "never"],
-		"space-before-blocks": [2, "always"],
-		"space-before-keywords": [2, "always"],
-		"no-console": 0,
-		"comma-dangle": 0,
-		"no-unexpected-multiline": 2,
-		"valid-jsdoc": 2
+		"strict": "off",
```
Comment (Kovensky) [inline .eslintrc]: I suppose this will come later?

Files that really need an exception (e.g. the hot reloading runtime) can always disable it with `/* eslint-disable strict */`.

File: .eslintrc [hunk 3]
```diff
@@ -8,29 +8,41 @@
...
+		"camelcase": "off",
+		"curly": "off",
+		"indent": ["error", "tab", { "SwitchCase": 1 }],
+		"eol-last": "warn",
+		"no-shadow": "off",
+		"no-redeclare": "error",
+		"no-extra-bind": "warn",
+		"no-empty": "off",
+		"no-process-exit": "warn",
+		"no-underscore-dangle": "off",
```
Comment (Kovensky) [inline .eslintrc]: This is already off by default

File: .eslintrc [hunk 4]
```diff
@@ -8,29 +8,41 @@
...
+		"no-use-before-define": "off",
```
Comment (Kovensky) [inline .eslintrc]: Also off by default

File: .eslintrc [hunk 5]
```diff
@@ -8,29 +8,41 @@
...
+		"no-undef": "error",
+		"no-unused-vars": "off",
```
Comment (Kovensky) [inline .eslintrc]: Probably should be turned on later 😄

File: .eslintrc [hunk 6]
```diff
@@ -8,29 +8,41 @@
...
+		"indent": ["error", "tab", { "SwitchCase": 1 }],
+		"eol-last": "warn",
+		"no-shadow": "off",
+		"no-redeclare": "error",
+		"no-extra-bind": "warn",
+		"no-empty": "off",
+		"no-process-exit": "warn",
+		"no-underscore-dangle": "off",
+		"no-use-before-define": "off",
+		"no-undef": "error",
```
Comment (Kovensky) [inline .eslintrc]: Already error by default (via `eslint:recommended`)

File: .eslintrc [hunk 7]
```diff
@@ -8,29 +8,41 @@
...
-		"no-loop-func": 1,
-		"space-before-function-paren": [2, "never"],
-		"space-before-blocks": [2, "always"],
-		"space-before-keywords": [2, "always"],
-		"no-console": 0,
-		"comma-dangle": 0,
-		"no-unexpected-multiline": 2,
-		"valid-jsdoc": 2
+		"strict": "off",
+		"camelcase": "off",
```
Comment (Kovensky) [inline .eslintrc]: This is already off by default

File: .eslintrc [hunk 8]
```diff
@@ -8,29 +8,41 @@
...
+		"curly": "off",
```
Comment (Kovensky) [inline .eslintrc]: Also off by default

**Comment by SpaceK33z** [comment]
@Kovensky, jup agree. I just spoke with sokra about this and I'm going to try to add the same js-beautify rules in eslint. That will also make it easier to use a linter in your text editor.

File: .eslintrc [hunk 9]
```diff
@@ -8,29 +8,41 @@
...
+		"indent": ["error", "tab", { "SwitchCase": 1 }],
+		"eol-last": "warn",
+		"no-shadow": "off",
+		"no-redeclare": "error",
```
Comment (Kovensky) [inline .eslintrc]: Already error by default (via `eslint:recommended`)

File: .eslintrc [hunk 10]
```diff
@@ -8,29 +8,41 @@
...
+		"no-extra-bind": "warn",
+		"no-empty": "off",
+		"no-process-exit": "warn",
+		"no-underscore-dangle": "off",
+		"no-use-before-define": "off",
+		"no-undef": "error",
+		"no-unused-vars": "off",
+		"consistent-return": "off",
```
Comment (Kovensky) [inline .eslintrc]: Also off by default

File: .eslintrc [hunk 11]
```diff
@@ -8,29 +8,41 @@
...
+		"no-inner-declarations": "warn",
```
Comment (Kovensky) [inline .eslintrc]: This is `error` by `eslint:recommended`; downgrading?

File: .eslintrc [hunk 12]
```diff
@@ -8,29 +8,41 @@
...
+				"try": {"after": true},
+				"else": {"after": true},
+				"throw": {"after": true},
+				"case": {"after": true},
+				"return": {"after": true},
+				"finally": {"after": true},
+				"do": {"after": true}
+			}
+		}],
+		"no-console": "off",
```
Comment (Kovensky) [inline .eslintrc]: Later on it'll probably be better to enable this and then specifically disable it on the files / blocks that are allowed to use `console`.

File: .eslintrc [hunk 13]
```diff
@@ -8,29 +8,41 @@
...
+		"comma-dangle": "off",
```
Comment (Kovensky) [inline .eslintrc]: Already off by default.

File: .eslintrc [hunk 14]
```diff
@@ -8,29 +8,41 @@
...
+		"no-unexpected-multiline": "error",
```
Comment (Kovensky) [inline .eslintrc]: Already error by default (via eslint:recommended).

File: .eslintrc (hunk 12 above)
Comment (SpaceK33z) [inline .eslintrc]: Good idea. We'll do that later sometime.

File: .eslintrc (hunk 2 above)
Comment (SpaceK33z) [inline .eslintrc]: We've started using strict mode for ES6 files, see [e.g. here](https://github.com/webpack/webpack/pull/3623/files).

Comment (Kovensky) [inline .eslintrc]: Yeah, would be nice if the linter could enforce that :)

The easiest way is to enable the rule, and then add disables to the specific files that are not written as strict.

**Comment by SpaceK33z** [comment]
I added rules that js-beautify already enforced to eslint. Note that eslint catched some errors that js-beautify didn't, these were fixed with `npm run beautify`. Long term plan is to remove js-beautify for eslint, but for now we can just run both next to each other.

File: lib/Dependency.js [hunk 15]
```diff
@@ -7,7 +7,7 @@ function Dependency() {
 }
 module.exports = Dependency;
 
-Dependency.prototype.isEqualResource = function( /* other */ ) {
```
Comment (sokra) [inline lib/Dependency.js]: It's sad that this has to be removed. It was intended as documentation of the available parameters

Comment (SpaceK33z) [inline lib/Dependency.js]: Understandable. The problem was that eslint and js-beautify don't agree with eachother about if there should be a space before and after the comment. js-beautify wants it with spaces, eslint without.

An alternative way of fixing this would be to add a comment add the top, `/* eslint-disable keyword-spacing */` (temporarily). After js-beautify is removed we can remove that line.

Comment (Kovensky) [inline lib/Dependency.js]: An alternative could be to just receive the parameter anyway and not use it. `no-unused-vars` is disabled here anyway, so there would be nothing to complain about it :cry:

Even if `no-unused-vars` were to be enabled, something like `{ "argsIgnorePattern": "^_" }` could be used to allow prefixing the argument name with a `_` to avoid the warning.

Comment (SpaceK33z) [inline lib/Dependency.js]: Ah nice idea, that may be the cleanest solution.

Comment (sokra) [inline lib/Dependency.js]: I wonder why `no-unused-vars` is disabled...

But the argsIgnorePattern seem to be a good option here.

Comment (SpaceK33z) [inline lib/Dependency.js]: I tried enabling it, but it resulted in so many errors that I thought it was best to fix this later.

File: package.json [hunk 16]
```diff
@@ -93,7 +93,6 @@
     "beautify-lint": "beautify-lint 'lib/**/*.js' 'hot/**/*.js' 'bin/**/*.js' 'benchmark/*.js' 'test/*.js'",
     "nsp": "nsp check --output summary",
     "cover": "node --harmony ./node_modules/.bin/istanbul cover -x '**/*.runtime.js' node_modules/mocha/bin/_mocha",
-    "beautify": "beautify-rewrite 'lib/**/*.js' 'hot/**/*.js' 'bin/**/*.js' 'benchmark/*.js' 'test/*.js'",
```
Comment (sokra) [inline package.json]: Why was this line removed?

Comment (SpaceK33z) [inline package.json]: I'm not sure, I think I meant to replace it with `npm run lint -- --fix`, but looks like I forgot that.
